voucher/views.py

```diff
@@ -1,3 +1,178 @@
+        else:
+            coupon_eligibility_choices = [
+                (match.id, match.product.course_run.first().course.title)
+                for match in coupon_eligibility_matches
+            ]
+            return render(
+                request,
+                "enroll.html",
+                context={
+                    "coupon_choices": coupon_eligibility_choices,
+                    **get_js_settings_context(self.request),
+                },
+            )
+
+    def post(self, request, *args, **kwargs):
+        voucher = get_current_voucher(self.request.user)
+        coupon_eligibility_id = request.POST["coupon_eligibility"][0]
+        coupon_eligibility = CouponEligibility.objects.get(id=coupon_eligibility_id)
+        voucher.coupon = coupon_eligibility.coupon
+        voucher.save()
+        enroll_url = f"/enroll?product_id={coupon_eligibility.product.id}&code={voucher.coupon.coupon_code}"
+        return redirect(enroll_url)
+
+
+@login_required
+def resubmit(request):
+    """
+    Prompt user to email voucher after failed voucher parsing
+    """
+    return render(request, "resubmit.html", context=get_js_settings_context(request))
+
+
+@login_required
+def redeemed(request):
+    """
+    Coupon has already been redeemed
+    """
+    return render(request, "redeemed.html", context=get_js_settings_context(request))
+
+
+def get_current_voucher(user):
+    """
+    Get the active voucher for a user
+    Args:
+        user (User): user to return voucher for
+
+    Returns:
+        voucher (Voucher): active voucher for user
+    """
+    return user.vouchers.order_by("uploaded").last()
+
+
+def match_courses_to_voucher(voucher):
+    """
+    Try to find an exact match for the course specified by a voucher
+
+    If no exact match is found, search for partial matches. Expand this function for more complex fuzzy matching
+
+    Args:
+        voucher(Voucher): a voucher to find courses for
+
+    Returns:
+        QuerySet: queryset of CourseRun objects
+    """
+    course_matches = CourseRun.objects.filter(
+        course__readable_id__exact=voucher.course_id_input,
+        course__title__iexact=voucher.course_title_input,
+        start_date__date=voucher.course_start_date_input,
+    )
+    if course_matches.count() > 1:
+        log.error("Found multiple exact CourseRun matches for voucher %s", voucher.id)
+        return CouponEligibility.objects.none()
+    elif not course_matches:
+        # Try partial matching
+        course_matches = CourseRun.objects.filter(
+            Q(course__readable_id__exact=voucher.course_id_input)
+            | Q(course__title__iexact=voucher.course_title_input)
+            | Q(start_date__date=voucher.course_start_date_input)
+        )
+    if not course_matches:
+        # No partial matches found
+        return CouponEligibility.objects.none()
+
+    # Check for valid coupon options and return choices
+    eligible_matches = CouponEligibility.objects.filter(
```

Comment:
This will return multiple coupons, and the coupons may be from another company (I created a batch of 20 single-use coupons from company MIT), and the dropdown list shows 20 choices.  

![Screen Shot 2019-05-29 at 12 55 04 PM](https://user-images.githubusercontent.com/187676/58576227-d5879c00-8211-11e9-893c-8a40a109d5f1.png)


If I select one of the choices and click 'Redeem', I get a `MultiValueDictKeyError` at /voucher/enroll on `coupon_eligibility`.

Only coupons from Boeing should be eligible.  `CouponPaymentVersion` has a `company` foreign key.  Could add another .env variable & setting, `VOUCHER_COMPANY_ID`, and filter by that.

Reply:
Might be better to revert to returning course runs here, and implement your earlier suggestion of using `get_valid_coupon_versions`, something like:

```python
    else:
        fuzzy_course_matches = CourseRun.objects.filter(
            Q(course__readable_id__exact=voucher.course_id_input)
            | Q(course__title__iexact=voucher.course_title_input)
            | Q(start_date__date=voucher.course_start_date_input)
        )

        return [
            match
            for match in fuzzy_course_matches
            if get_valid_coupon_versions(
                match.product.first(),
                voucher.user,
                auto_only=True,
                code=None,
                full_discount=True,
                company=Company.objects.get(id=settings.VOUCHER_COMPANY_ID),
            )
            is not None
        ]
```

Reply:
Another option would be to modify `get_valid_coupon_versions` to accept a list of products instead of one product as an input argument, though that will require refactoring of existing functionality that depends on it.

Reply:
@mbertrand What if we did something like this:
```
valid_coupons = [
        get_valid_coupon_versions(
            match.product.first(),
            voucher.user,
            full_discount=True,
        ).first() for match in course_matches
    ]
eligible_choices = [
        ((course_matches[i].product, valid_coupons[i].coupon.id), course_matches[i].course.title)
        for i in range(len(course_matches))
        if valid_coupons[i] is not None and course_matches[i].product is not None
    ]
```

I'll add the company back in as a separate PR after this is working. I know we want to get this in ASAP.

Reply:
Still need to filter out coupons attached to any existing voucher, in addition to filtering by company.  I'd prefer both to be in this PR since they're pretty crucial to this working as intended.

Reply:
Ok. I'll add them in tomorrow. I'll set the VOUCHER_COMPANY_ID to boeing by default for now, but we will have to parse it out of the voucher at some point.